File: clients/python/sensocto/call.py
"""Call session implementation."""

import asyncio
import logging
from typing import Any, Callable, Dict, List, Optional

from sensocto.errors import DisconnectedError, SensoctoError
from sensocto.models import (
    CallEvent,
    CallParticipant,
    CallEndedEvent,
    MediaEventReceived,
    ParticipantAudioChangedEvent,
    ParticipantJoinedEvent,
    ParticipantLeftEvent,
    ParticipantVideoChangedEvent,
    QualityChangedEvent,
)
from sensocto.socket import PhoenixSocket

logger = logging.getLogger(__name__)


class CallSession:
    """Session for video/voice communication."""

    def __init__(
        self,
        socket: PhoenixSocket,
        topic: str,
        room_id: str,
        user_id: str,
        ice_servers: List[Any],
    ):
        """
        Creates a new call session.

        Args:
            socket: The Phoenix socket.
            topic: The channel topic.
            room_id: The room ID.
            user_id: The user ID.
            ice_servers: ICE servers for WebRTC.
        """
        self._socket = socket
        self._topic = topic
        self._room_id = room_id
        self._user_id = user_id
        self._ice_servers = ice_servers
        self._joined = False
        self._in_call = False
        self._endpoint_id: Optional[str] = None
        self._event_handlers: List[Callable[[CallEvent], None]] = []

        # Register event handlers
        self._setup_event_handlers()

    @property
    def room_id(self) -> str:
        """Returns the room ID."""
        return self._room_id

    @property
    def user_id(self) -> str:
        """Returns the user ID."""
        return self._user_id

    @property
    def in_call(self) -> bool:
        """Returns whether the user is in the call."""
        return self._in_call

    @property
    def endpoint_id(self) -> Optional[str]:
        """Returns the endpoint ID."""
        return self._endpoint_id

    @property
    def ice_servers(self) -> List[Any]:
        """Returns the ICE servers."""
        return self._ice_servers

    def on_event(self, handler: Callable[[CallEvent], None]) -> None:
        """
        Registers an event handler.

        Args:
            handler: Callback function called when events are received.
        """
        self._event_handlers.append(handler)

    async def join_channel(self, join_params: Dict[str, Any]) -> Dict[str, Any]:
        """
        Joins the call channel.

        Args:
            join_params: Parameters for joining the channel.

        Returns:
            The join response from the server.
        """
        reply = await self._socket.send(self._topic, "phx_join", join_params)

        if reply.is_ok:
            self._joined = True
            # Extract ICE servers from response
            if "ice_servers" in reply.response:
                self._ice_servers = reply.response["ice_servers"]
            logger.info(f"Joined call channel: {self._topic}")
            return reply.response
        else:
            raise SensoctoError(f"Failed to join channel: {reply.response}")

    async def leave_channel(self) -> None:
        """Leaves the call channel."""
        if not self._joined:
            return

        if self._in_call:
            await self.leave_call()

        await self._socket.send(self._topic, "phx_leave", {})
        self._joined = False
        logger.info(f"Left call channel: {self._topic}")

    async def join_call(self) -> Dict[str, Any]:
        """
        Joins the actual call.

        Returns:
            The join response with endpoint_id and participants.
        """
        if not self._joined:
            raise SensoctoError("Channel not joined")

        reply = await self._socket.send(self._topic, "join_call", {})

        if reply.is_ok:
            self._in_call = True
            self._endpoint_id = reply.response.get("endpoint_id")
            logger.info(f"Joined call with endpoint: {self._endpoint_id}")
            return reply.response
        else:
            raise SensoctoError(f"Failed to join call: {reply.response}")

    async def leave_call(self) -> None:
        """Leaves the call."""
        if not self._in_call:
            return

        await self._socket.send(self._topic, "leave_call", {})
        self._in_call = False
        self._endpoint_id = None
        logger.info("Left call")

    async def send_media_event(self, data: Any) -> None:
        """
        Sends a media event (SDP offer/answer, ICE candidate).

        Args:
            data: The media event data.
        """
        if not self._in_call:
            raise SensoctoError("Not in call")

        await self._socket.send_no_reply(self._topic, "media_event", {"data": data})

    async def toggle_audio(self, enabled: bool) -> None:
        """
        Toggles the local audio state.

        Args:
            enabled: Whether audio should be enabled.
        """
        if not self._in_call:
            raise SensoctoError("Not in call")

        await self._socket.send(self._topic, "toggle_audio", {"enabled": enabled})

    async def toggle_video(self, enabled: bool) -> None:
        """
        Toggles the local video state.

        Args:
            enabled: Whether video should be enabled.
        """
        if not self._in_call:
            raise SensoctoError("Not in call")

        await self._socket.send(self._topic, "toggle_video", {"enabled": enabled})

    async def set_quality(self, quality: str) -> None:
        """
        Sets the video quality.

        Args:
            quality: Quality level ("high", "medium", "low", or "auto").
        """
        if not self._in_call:
            raise SensoctoError("Not in call")

        await self._socket.send(self._topic, "set_quality", {"quality": quality})

    async def get_participants(self) -> Dict[str, CallParticipant]:
        """
        Gets the current participants.

        Returns:
            Dictionary mapping user_id to CallParticipant.
        """
        reply = await self._socket.send(self._topic, "get_participants", {})

        if reply.is_ok:
            participants = {}
            for user_id, data in reply.response.get("participants", {}).items():
                participants[user_id] = CallParticipant(
                    user_id=data.get("user_id", user_id),
                    endpoint_id=data.get("endpoint_id", ""),
                    user_info=data.get("user_info", {}),
                    joined_at=data.get("joined_at"),
                    audio_enabled=data.get("audio_enabled", False),
                    video_enabled=data.get("video_enabled", False),
                )
            return participants

        return {}

    def _setup_event_handlers(self) -> None:
        """Sets up event handlers for call events."""
        self._socket.on(self._topic, "participant_joined", self._on_participant_joined)
        self._socket.on(self._topic, "participant_left", self._on_participant_left)
        self._socket.on(self._topic, "media_event", self._on_media_event)
        self._socket.on(self._topic, "participant_audio_changed", self._on_audio_changed)
        self._socket.on(self._topic, "participant_video_changed", self._on_video_changed)
        self._socket.on(self._topic, "quality_changed", self._on_quality_changed)
        self._socket.on(self._topic, "call_ended", self._on_call_ended)

    def _dispatch_event(self, event: CallEvent) -> None:
        """Dispatches an event to all handlers."""
        for handler in self._event_handlers:
            try:
                handler(event)
            except Exception as e:
                logger.error(f"Error in event handler: {e}")

    def _on_participant_joined(self, payload: Dict[str, Any]) -> None:
        """Handles participant joined event."""
        participant = CallParticipant(
            user_id=payload.get("user_id", ""),
            endpoint_id=payload.get("endpoint_id", ""),
            user_info=payload.get("user_info", {}),
            joined_at=payload.get("joined_at"),
            audio_enabled=payload.get("audio_enabled", False),
            video_enabled=payload.get("video_enabled", False),
        )
        self._dispatch_event(ParticipantJoinedEvent(participant=participant))

    def _on_participant_left(self, payload: Dict[str, Any]) -> None:
        """Handles participant left event."""
        self._dispatch_event(ParticipantLeftEvent(
            user_id=payload.get("user_id", ""),
            crashed=payload.get("crashed", False),
        ))

    def _on_media_event(self, payload: Dict[str, Any]) -> None:
        """Handles media event."""
        self._dispatch_event(MediaEventReceived(data=payload.get("data")))

    def _on_audio_changed(self, payload: Dict[str, Any]) -> None:
        """Handles audio changed event."""
        self._dispatch_event(ParticipantAudioChangedEvent(
            user_id=payload.get("user_id", ""),
            enabled=payload.get("audio_enabled", False),
        ))

    def _on_video_changed(self, payload: Dict[str, Any]) -> None:
        """Handles video changed event."""
        self._dispatch_event(ParticipantVideoChangedEvent(
            user_id=payload.get("user_id", ""),
            enabled=payload.get("video_enabled", False),
        ))

    def _on_quality_changed(self, payload: Dict[str, Any]) -> None:
        """Handles quality changed event."""
        self._dispatch_event(QualityChangedEvent(quality=payload.get("quality", "")))

    def _on_call_ended(self, payload: Dict[str, Any]) -> None:
        """Handles call ended event."""
        self._in_call = False
        self._dispatch_event(CallEndedEvent())

    async def close(self) -> None:
        """Closes the call session."""
        await self.leave_channel()

File: clients/threejs/src/call.ts
/**
 * Call session implementation for video/voice communication.
 * @module call
 */

import { PhoenixChannel } from "./channel.js";
import { SensoctoError } from "./errors.js";
import { type AnyCallEvent, type CallParticipant, type IceServer } from "./models.js";

/**
 * Callback for call events.
 */
export type CallEventHandler = (event: AnyCallEvent) => void;

/**
 * Session for video/voice communication.
 *
 * Manages WebRTC signaling through Phoenix channels for
 * real-time video and voice calls.
 *
 * @example
 * ```typescript
 * const session = await client.joinCall("room-123", "user-456");
 *
 * session.onEvent((event) => {
 *   switch (event.type) {
 *     case "participant_joined":
 *       console.log("Joined:", event.participant);
 *       break;
 *     case "media_event":
 *       // Handle WebRTC signaling
 *       handleMediaEvent(event.data);
 *       break;
 *   }
 * });
 *
 * // Join the actual call
 * await session.joinCall();
 *
 * // Send WebRTC signaling data
 * await session.sendMediaEvent(sdpOffer);
 * ```
 */
export class CallSession {
  private readonly channel: PhoenixChannel;
  private readonly _roomId: string;
  private readonly _userId: string;
  private _iceServers: IceServer[];
  private _inCall = false;
  private _endpointId: string | null = null;
  private eventHandlers: CallEventHandler[] = [];

  /**
   * Creates a new call session.
   *
   * @param channel - The Phoenix channel for this call
   * @param roomId - The room ID
   * @param userId - The user ID
   * @param iceServers - ICE servers for WebRTC
   */
  constructor(
    channel: PhoenixChannel,
    roomId: string,
    userId: string,
    iceServers: IceServer[] = []
  ) {
    this.channel = channel;
    this._roomId = roomId;
    this._userId = userId;
    this._iceServers = iceServers;

    this.setupEventHandlers();
  }

  /**
   * Returns the room ID.
   */
  get roomId(): string {
    return this._roomId;
  }

  /**
   * Returns the user ID.
   */
  get userId(): string {
    return this._userId;
  }

  /**
   * Returns whether the user is in the call.
   */
  get inCall(): boolean {
    return this._inCall;
  }

  /**
   * Returns the endpoint ID.
   */
  get endpointId(): string | null {
    return this._endpointId;
  }

  /**
   * Returns the ICE servers.
   */
  get iceServers(): IceServer[] {
    return this._iceServers;
  }

  /**
   * Returns the ICE servers in RTCPeerConnection format.
   */
  get rtcIceServers(): RTCIceServer[] {
    return this._iceServers.map((server) => {
      const result: RTCIceServer = { urls: server.urls };
      if (server.username !== undefined) {
        result.username = server.username;
      }
      if (server.credential !== undefined) {
        result.credential = server.credential;
      }
      return result;
    });
  }

  /**
   * Registers an event handler.
   *
   * @param handler - Callback function called when events are received
   * @returns A function to unsubscribe the handler
   */
  onEvent(handler: CallEventHandler): () => void {
    this.eventHandlers.push(handler);

    return () => {
      const index = this.eventHandlers.indexOf(handler);
      if (index !== -1) {
        this.eventHandlers.splice(index, 1);
      }
    };
  }

  /**
   * Joins the actual call.
   *
   * @returns The join response with endpoint_id and participants
   * @throws {SensoctoError} If not connected to channel
   */
  async joinCall(): Promise<{
    endpointId: string;
    participants: Record<string, CallParticipant>;
  }> {
    if (!this.channel.isJoined) {
      throw new SensoctoError("Channel not joined");
    }

    const reply = await this.channel.push("join_call", {});

    if (reply.status === "ok") {
      this._inCall = true;
      this._endpointId = (reply.response["endpoint_id"] as string) ?? null;

      const rawParticipants = (reply.response["participants"] as Record<string, unknown>) ?? {};
      const participants: Record<string, CallParticipant> = {};

      for (const [odUserId, data] of Object.entries(rawParticipants)) {
        const p = data as Record<string, unknown>;
        const participant: CallParticipant = {
          userId: (p["user_id"] as string) ?? odUserId,
          endpointId: (p["endpoint_id"] as string) ?? "",
          userInfo: (p["user_info"] as Record<string, unknown>) ?? {},
          audioEnabled: (p["audio_enabled"] as boolean) ?? false,
          videoEnabled: (p["video_enabled"] as boolean) ?? false,
        };
        if (typeof p["joined_at"] === "string") {
          participant.joinedAt = p["joined_at"];
        }
        participants[odUserId] = participant;
      }

      return { endpointId: this._endpointId ?? "", participants };
    }

    throw new SensoctoError(`Failed to join call: ${JSON.stringify(reply.response)}`);
  }

  /**
   * Leaves the call.
   */
  async leaveCall(): Promise<void> {
    if (!this._inCall) {
      return;
    }

    await this.channel.push("leave_call", {});
    this._inCall = false;
    this._endpointId = null;
  }

  /**
   * Sends a media event (SDP offer/answer, ICE candidate).
   *
   * @param data - The media event data
   * @throws {SensoctoError} If not in call
   */
  async sendMediaEvent(data: unknown): Promise<void> {
    if (!this._inCall) {
      throw new SensoctoError("Not in call");
    }

    await this.channel.pushNoReply("media_event", { data });
  }

  /**
   * Toggles the local audio state.
   *
   * @param enabled - Whether audio should be enabled
   * @throws {SensoctoError} If not in call
   */
  async toggleAudio(enabled: boolean): Promise<void> {
    if (!this._inCall) {
      throw new SensoctoError("Not in call");
    }

    await this.channel.push("toggle_audio", { enabled });
  }

  /**
   * Toggles the local video state.
   *
   * @param enabled - Whether video should be enabled
   * @throws {SensoctoError} If not in call
   */
  async toggleVideo(enabled: boolean): Promise<void> {
    if (!this._inCall) {
      throw new SensoctoError("Not in call");
    }

    await this.channel.push("toggle_video", { enabled });
  }

  /**
   * Sets the video quality.
   *
   * @param quality - Quality level ("high", "medium", "low", or "auto")
   * @throws {SensoctoError} If not in call
   */
  async setQuality(quality: "high" | "medium" | "low" | "auto"): Promise<void> {
    if (!this._inCall) {
      throw new SensoctoError("Not in call");
    }

    await this.channel.push("set_quality", { quality });
  }

  /**
   * Gets the current participants.
   *
   * @returns Dictionary mapping user_id to CallParticipant
   */
  async getParticipants(): Promise<Record<string, CallParticipant>> {
    const reply = await this.channel.push("get_participants", {});

    if (reply.status === "ok") {
      const rawParticipants = (reply.response["participants"] as Record<string, unknown>) ?? {};
      const participants: Record<string, CallParticipant> = {};

      for (const [odUserId, data] of Object.entries(rawParticipants)) {
        const p = data as Record<string, unknown>;
        const participant: CallParticipant = {
          userId: (p["user_id"] as string) ?? odUserId,
          endpointId: (p["endpoint_id"] as string) ?? "",
          userInfo: (p["user_info"] as Record<string, unknown>) ?? {},
          audioEnabled: (p["audio_enabled"] as boolean) ?? false,
          videoEnabled: (p["video_enabled"] as boolean) ?? false,
        };
        if (typeof p["joined_at"] === "string") {
          participant.joinedAt = p["joined_at"];
        }
        participants[odUserId] = participant;
      }

      return participants;
    }

    return {};
  }

  /**
   * Closes the call session.
   */
  async close(): Promise<void> {
    if (this._inCall) {
      await this.leaveCall();
    }

    await this.channel.leave();
    this.eventHandlers = [];
  }

  private setupEventHandlers(): void {
    this.channel.on("participant_joined", (payload) => {
      const participant: CallParticipant = {
        userId: (payload["user_id"] as string) ?? "",
        endpointId: (payload["endpoint_id"] as string) ?? "",
        userInfo: (payload["user_info"] as Record<string, unknown>) ?? {},
        audioEnabled: (payload["audio_enabled"] as boolean) ?? false,
        videoEnabled: (payload["video_enabled"] as boolean) ?? false,
      };
      if (typeof payload["joined_at"] === "string") {
        participant.joinedAt = payload["joined_at"];
      }
      this.dispatchEvent({
        type: "participant_joined",
        participant,
      });
    });

    this.channel.on("participant_left", (payload) => {
      this.dispatchEvent({
        type: "participant_left",
        userId: (payload["user_id"] as string) ?? "",
        crashed: (payload["crashed"] as boolean) ?? false,
      });
    });

    this.channel.on("media_event", (payload) => {
      this.dispatchEvent({
        type: "media_event",
        data: payload["data"],
      });
    });

    this.channel.on("participant_audio_changed", (payload) => {
      this.dispatchEvent({
        type: "participant_audio_changed",
        userId: (payload["user_id"] as string) ?? "",
        enabled: (payload["audio_enabled"] as boolean) ?? false,
      });
    });

    this.channel.on("participant_video_changed", (payload) => {
      this.dispatchEvent({
        type: "participant_video_changed",
        userId: (payload["user_id"] as string) ?? "",
        enabled: (payload["video_enabled"] as boolean) ?? false,
      });
    });

    this.channel.on("quality_changed", (payload) => {
      this.dispatchEvent({
        type: "quality_changed",
        quality: (payload["quality"] as string) ?? "",
      });
    });

    this.channel.on("call_ended", () => {
      this._inCall = false;
      this.dispatchEvent({ type: "call_ended" });
    });
  }

  private dispatchEvent(event: AnyCallEvent): void {
    for (const handler of this.eventHandlers) {
      try {
        handler(event);
      } catch (error) {
        console.error("Error in call event handler:", error);
      }
    }
  }
}
